src/fidesops/util/saas_util.py

```diff
@@ -165,3 +167,21 @@ def format_body(
         raise FidesopsException(f"Unsupported Content-Type: {content_type}")
 
     return headers, output
+
+
+def assign_placeholders(value: str, param_values: Dict[str, Any]) -> Optional[str]:
+    """
+    Finds all the placeholders (indicated by <>) in the passed in value
+    and replaces them with the actual param values
+
+    Returns None if any of the placeholders cannot be found in the param_values
+    """
+    if value and isinstance(value, str):
+        placeholders = re.findall("<(.+?)>", value)
+        for placeholder in placeholders:
+            placeholder_value = param_values.get(placeholder)
+            if placeholder_value:
+                value = value.replace(f"<{placeholder}>", str(placeholder_value))
+            else:
+                return None
```

Comment:
This function will still have mutated `value` if the second (or later) `placeholder` has no `placeholder_value` in `param_values`. Is this intended behavior? If not, we should duplicate `value`, mutate our duplicate, and return either that new string or `None`. We should also add test coverage for that case.

Separately, does every instance where `assign_placeholders` is called currently handle the possibility of `None` being returned? It might be more intuitive to instead `raise` a new custom exception.

Reply:
Ok, but the first issue still stands - if multiple `placeholder`s exist, but only a single `placeholder_value`, then `value` itself will still have been mutated when the function returns `None`. It's important that we add test coverage for this helper, so we can assert that it behaves as expected.